system/mp/mpcomm.cpp

```diff
@@ -1725,136 +1828,130 @@ class CMPPacketReader: public ISocketSelectNotify, public CInterface
-                    }
-                    else
-                        sock->read(&hdr,sizeof(hdr));
-                    if (hdr.version/0x100 != MP_PROTOCOL_VERSION/0x100) {
+                    remaining = sizeof(PacketHeader);
+                    gotPacketHdr = false;
+                    activemsg = new CMessageBuffer(remaining);
+                    activeptr = (byte *)activemsg->bufferBase();
+                }
+                size32_t szRead;
+                if (!gotPacketHdr)
+                {
+                    CCycleTimer timer;
+                    sock->readtms(activeptr, 0, remaining, szRead, timer.remainingMs(60000));
+                    remaining -= szRead;
+                    activeptr += szRead;
+                    if (remaining) // only possible if blocked.
+                        return false; // wait for next notification
+
+                    PacketHeader &hdr = *(PacketHeader *)activemsg->bufferBase();
+                    if (hdr.version/0x100 != MP_PROTOCOL_VERSION/0x100)
+                    {
                         // TBD IPV6 here
                         SocketEndpoint ep;
                         sock->getPeerEndpoint(ep);
                         IMP_Exception *e=new CMPException(MPERR_protocol_version_mismatch,ep);
                         throw e;
                     }
-                    if (sizeavail<=sizeof(hdr))
-                        sizeavail = sock->avail_read(); 
-                    else
-                        sizeavail -= sizeof(hdr);
-#ifdef _FULLTRACE
+                    hdr.setMessageFields(*activemsg);
+    #ifdef _FULLTRACE
                     StringBuffer ep1;
                     StringBuffer ep2;
                     LOG(MCdebugInfo, "MP: ReadPacket(sender=%s,target=%s,tag=%d,replytag=%d,size=%d)",hdr.sender.getEndpointHostText(ep1).str(),hdr.target.getEndpointHostText(ep2).str(),hdr.tag,hdr.replytag,hdr.size);
-#endif
+    #endif
                     remaining = hdr.size-sizeof(hdr);
-                    activemsg = new CMessageBuffer(remaining); // will get from low level IO at some stage
-                    activeptr = (byte *)activemsg->reserveTruncate(remaining);
-                    hdr.setMessageFields(*activemsg);
+                    activeptr = (byte *)activemsg->clear().reserveTruncate(remaining);
+                    gotPacketHdr = true;
                 }
-                
-                size32_t toread = sizeavail;
-                if (toread>remaining)
-                    toread = remaining;
-                if (toread) {
-                    sock->read(activeptr,toread);
-                    remaining -= toread;
-                    sizeavail -= toread;
-                    activeptr += toread;
+
+                if (remaining)
+                {
+                    sock->readtms(activeptr, 0, remaining, szRead, WAIT_FOREVER);
```

Comment:
WAIT_FOREVER here ?

Reply:
it was made that, in "sympathy" to the way it was previously coded:
```
sock->read(&hdr,sizeof(hdr));
```

but we could rationalize at the same time as this PR too..
To be conservative, could make it a reasonably long time JIC e.g. 15 minutes - of course it should never take that long, but... don't want to change this too much at same time

Reply:
I missed the main point here... in above reply.
It is readtms with min_size = 0, the timeout is basically irrelevant - it can't hit the timeout, because if blocked it will bail out.
There's a comment in CSocket::readtms:
```
     * NB: timeout is meaningless if min_size is 0
```